### ISSUE TITLE
Weight Decay with BERT 

### ISSUE THREAD

**Issue by Steven-Palayew** [amaiya/ktrain#424]
Hello, 

For BERT and BERT based transformers, I've heard when applying weight decay one should set the amount of weight decay for biases and LayerNorm weights to 0. How would this be done with ktrain? 

Thanks in advance!

**Comment by amaiya** [comment]
The default should already be zero.  If you want to use a weight decay greater than zero, there is the `learner.set_weight_decay` method, which essentially switches out the default Adam optimizer and uses the AdamW optimizer and sets the weight decay accordingly.   Doing this is essentially the same as compiling the `tf.keras` model to use the AdamWeightDecay optimizer as a custom optimizer from the `transformers` library.    For more information, refer to the [AdamW documentation](https://huggingface.co/docs/transformers/main_classes/optimizer_schedules#transformers.AdamWeightDecay) and [this post](https://discuss.huggingface.co/t/does-the-default-weight-decay-of-0-0-in-transformers-adamw-make-sense/1180).

**Comment by Steven-Palayew** [comment]
> The default should already be zero. If you want to use a weight decay greater than zero, there is the `learner.set_weight_decay` method, which essentially switches out the default Adam optimizer and uses the AdamW optimizer and sets the weight decay accordingly. Doing this is essentially the same as compiling the `tf.keras` model to use the AdamWeightDecay optimizer as a custom optimizer from the `transformers` library. For more information, refer to the [AdamW documentation](https://huggingface.co/docs/transformers/main_classes/optimizer_schedules#transformers.AdamWeightDecay) and [this post](https://discuss.huggingface.co/t/does-the-default-weight-decay-of-0-0-in-transformers-adamw-make-sense/1180).

Hi, to clarify I mean that I want to set a global weight decay of around 0.01, but I want to exclude these parameters, as this seems to be best practice.

**Comment by amaiya** [comment]
Hi:  I misread your initial question, but the answer turns out to be the same:  this is already the default in ktrain when using a weight decay > 0:
```python
DEFAULT_WD = 0.01
def get_default_optimizer(lr=0.001, wd=DEFAULT_WD):
    from .lroptimize.optimization import AdamWeightDecay
    opt = AdamWeightDecay(learning_rate=lr, 
                         weight_decay_rate=wd, 
                         beta_1=0.9,
                         beta_2=0.999,
                         epsilon=1e-6,
                         exclude_from_weight_decay=['layer_norm', 'bias'])
```
